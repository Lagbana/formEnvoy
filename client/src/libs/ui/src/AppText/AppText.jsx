/* HTML-enabled text block
=============================== */

import React from 'react';
import { View, Text, StyleSheet, Platform, TextStyle } from 'react-native';
import HTML from 'react-native-render-html';
import { useInAppBrowser } from '../hooks';
import { Typography, Colors } from 'libs/ui/src/WebButton/node_modules/libs/ui/src/WebLoading/node_modules/colibri/styles';
import { isNil, isString } from 'lodash';

export interface TagStyle {
	[key: string]: TextStyle;
}

export interface AppTextProps {
	fontStyle?: TextStyle;
	tagsStyles?: TagStyle;
	style?: TextStyle;
	classesStyles?: HTML.StylesDictionary;
	onLinkPress?: (url: string) => void;
}

export const AppText: React.FC<AppTextProps> = ({
  children,
  fontStyle,
  tagsStyles,
  classesStyles,
  onLinkPress,
	style
}) => {
  const { openLink } = useInAppBrowser();

  // styles

  const styles = StyleSheet.create({
    container: {
      marginBottom: 20,
    },
  });

  const baseFontStyle = {
    ...Typography.bodyLarge,
    color: Colors.text,
    ...style,
    ...fontStyle,
  };

  const baseTagStyles = {
    b: { fontFamily: Typography.families.bold, fontWeight: null },
    strong: { fontFamily: Typography.families.bold, fontWeight: null },
    a: { color: Colors.blue, textDecorationLine: 'none' },
    ...tagsStyles,
  };

	if (isNil(children) || children === '') { return null; }

	if (Platform.OS === 'web') {

		const content = (isString(children)) ? <div dangerouslySetInnerHTML={{ __html: children }} /> : children;

		return (
			<Text style={[styles.container, baseFontStyle, style]}>
				{content}
			</Text>
		);

	} else {

		const content = (isString(children)) ? <HTML
					html={children}
					baseFontStyle={baseFontStyle}
					tagsStyles={baseTagStyles}
					classesStyles={classesStyles}
					onLinkPress={(e, href) => {
						openLink(href);
						if(onLinkPress) { onLinkPress(href); }
					}}
				/>: children;

		return (
			<View style={[styles.container, style]}>
				{content}
			</View>
		)
	}
};
