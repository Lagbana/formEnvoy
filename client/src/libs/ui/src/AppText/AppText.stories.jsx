import React from 'react'
import { storiesOf } from '@storybook/react-native'
import { AppText } from './AppText'

storiesOf('AppText', module).add('Default', () => (
  <AppText onLinkPress={(string: 'test') => null} children={'Hello World!'} />
))
