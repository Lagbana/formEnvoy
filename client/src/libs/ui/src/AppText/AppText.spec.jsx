import React from 'react'
import { shallow, configure } from 'enzyme'
import toJson from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16'
import { AppText } from './AppText';
import { Text } from 'react-native';

configure({ adapter: new Adapter() })

test('<AppText /> renders empty string correctly', () => {
  const component = shallow(
    <AppText children={''} />
  )
  expect(toJson(component)).toMatchSnapshot()
})

test('<AppText /> renders html correctly', () => {
  const component = shallow(
    <AppText children={'<bold>Some bold text</bold>'} />
  )
  expect(toJson(component)).toMatchSnapshot()
})

test('<AppText /> renders null correctly', () => {
  const component = shallow(
    <AppText children={null} />
  )
  expect(toJson(component)).toMatchSnapshot()
})

test('<AppText /> renders undefined correctly', () => {
  const component = shallow(
    <AppText children={undefined} />
  )
  expect(toJson(component)).toMatchSnapshot()
})

test('<AppText /> renders child components correctly correctly', () => {
  const component = shallow(
    <AppText children={<Text>My child</Text>} />
  )
  expect(toJson(component)).toMatchSnapshot()
})

