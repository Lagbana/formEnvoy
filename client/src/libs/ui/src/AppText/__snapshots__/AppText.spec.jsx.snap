// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppText /> renders child components correctly correctly 1`] = `
<View
  style={
    Array [
      Object {
        "marginBottom": 20,
      },
      undefined,
    ]
  }
>
  <Text>
    My child
  </Text>
</View>
`;

exports[`<AppText /> renders empty string correctly 1`] = `""`;

exports[`<AppText /> renders html correctly 1`] = `
<View
  style={
    Array [
      Object {
        "marginBottom": 20,
      },
      undefined,
    ]
  }
>
  <HTML
    allowFontScaling={true}
    baseFontStyle={
      Object {
        "color": "#3E3F4B",
        "fontFamily": "ObjektivMk1-Regular",
        "fontSize": 20,
        "lineHeight": 28,
        "paddingTop": 5,
      }
    }
    classesStyles={Object {}}
    debug={false}
    decodeEntities={true}
    emSize={14}
    html="<bold>Some bold text</bold>"
    ignoredStyles={Array []}
    ignoredTags={
      Array [
        "head",
        "script",
        "audio",
        "video",
        "track",
        "embed",
        "object",
        "param",
        "source",
        "canvas",
        "noscript",
        "caption",
        "col",
        "colgroup",
        "table",
        "tbody",
        "td",
        "tfoot",
        "th",
        "thead",
        "tr",
        "button",
        "datalist",
        "fieldset",
        "form",
        "input",
        "label",
        "legend",
        "meter",
        "optgroup",
        "option",
        "output",
        "progress",
        "select",
        "textarea",
        "details",
        "dialog",
        "menu",
        "menuitem",
        "summary",
      ]
    }
    imagesMaxWidth={750}
    onLinkPress={[Function]}
    ptSize={1.3}
    renderers={
      Object {
        "a": [Function],
        "br": [Function],
        "iframe": [Function],
        "img": [Function],
        "ol": [Function],
        "pre": [Function],
        "textwrapper": [Function],
        "ul": [Function],
      }
    }
    staticContentMaxWidth={750}
    tagsStyles={
      Object {
        "a": Object {
          "color": "#3485D7",
          "textDecorationLine": "none",
        },
        "b": Object {
          "fontFamily": "ObjektivMk1-Bold",
          "fontWeight": null,
        },
        "strong": Object {
          "fontFamily": "ObjektivMk1-Bold",
          "fontWeight": null,
        },
      }
    }
    textSelectable={false}
  />
</View>
`;

exports[`<AppText /> renders null correctly 1`] = `""`;

exports[`<AppText /> renders undefined correctly 1`] = `""`;
